rx/scheduler/trampolinescheduler.py

```diff
@@ -0,0 +1,164 @@
+        """Creates a scheduler that schedules work as soon as possible
+        on the current thread."""
+
+        self._lock = Lock()
+        self._local: MutableMapping[Thread, _Local] = WeakKeyDictionary()
+
+    def schedule(self,
+                 action: typing.ScheduledAction,
+                 state: Optional[typing.TState] = None
+                 ) -> typing.Disposable:
+        """Schedules an action to be executed.
+
+        Args:
+            action: Action to be executed.
+            state: [Optional] state to be given to the action function.
+
+        Returns:
+            The disposable object used to cancel the scheduled action
+            (best effort).
+        """
+
+        return self.schedule_absolute(self.now, action, state=state)
+
+    def schedule_relative(self,
+                          duetime: typing.RelativeTime,
+                          action: typing.ScheduledAction,
+                          state: Optional[typing.TState] = None
+                          ) -> typing.Disposable:
+        """Schedules an action to be executed after duetime.
+
+        Args:
+            duetime: Relative time after which to execute the action.
+            action: Action to be executed.
+            state: [Optional] state to be given to the action function.
+
+        Returns:
+            The disposable object used to cancel the scheduled action
+            (best effort).
+        """
+
+        duetime = max(DELTA_ZERO, self.to_timedelta(duetime))
+        return self.schedule_absolute(self.now + duetime, action, state=state)
+
+    def schedule_absolute(self,
+                          duetime: typing.AbsoluteTime,
+                          action: typing.ScheduledAction,
+                          state: Optional[typing.TState] = None
+                          ) -> typing.Disposable:
+        """Schedules an action to be executed at duetime.
+
+        Args:
+            duetime: Absolute time at which to execute the action.
+            action: Action to be executed.
+            state: [Optional] state to be given to the action function.
+
+        Returns:
+            The disposable object used to cancel the scheduled action
+            (best effort).
+        """
+
+        duetime = self.to_datetime(duetime)
+
+        if duetime > self.now:
+            log.warning('Do not schedule blocking work!')
+
+        si: ScheduledItem[typing.TState] = ScheduledItem(self, state, action, duetime)
+
+        local = self._get_local()
+        local.queue.enqueue(si)
+        if local.idle:
+            local.idle = False
+            try:
+                _Trampoline.run(local.queue)
+            finally:
+                local.idle = True
+                local.queue.clear()
+
+        return si.disposable
+
+    def _get_local(self):
+        thread = current_thread()
+        local = self._local.get(thread)
+        if local is None:
+            local = _Local()
+            self._local[thread] = local
```

Comment:
Alternatively, this could be simplified by using [`set_default`](https://docs.python.org/3.7/library/stdtypes.html#dict.setdefault) method from dict, i.e.:
```python
local = self._local.setdefault(thread, _Local())
```

Reply:
Also should writing to self._local weak dict be protected by a lock?